### PR TITLE
Initial base files

### PR DESCRIPTION
**Pull request by majestrate**
Contains:

* boilerplate for all of the i2p specific crypto parts minus aes

* some boilerplate for reseed and main function

* .gitignore

* some initial datastructures

### CODE REVIEW

File: lib/config/router.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package config
+
+// router.config options
+type RouterConfig struct {
+  NetDbDir string
```
Comment (hkparker) [inline lib/config/router.go]: Is the goal to be compatible with I2P's directory structure (so either application could be ran from the same configuration directory)?  That would be great, but if not perhaps we could explore alternative storage (sqlite?)

File: lib/crypto/dsa_test.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package crypto
+
+import (
+  "crypto/dsa"
+  "crypto/rand"
+  "testing"
+)
+
+
+func TestDSA(t *testing.T) {
```
Comment (hkparker) [inline lib/crypto/dsa_test.go]: Do you have an option regarding go's built in testing tools vs. frameworks like ginkgo?  I generally find BDD convenient but I'm not married to it.

File: lib/crypto/ed25519.go [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package crypto
+
+
```
Comment (hkparker) [inline lib/crypto/ed25519.go]: Could look into https://github.com/agl/ed25519

File: lib/netdb/doc.go [hunk 4]
```diff
@@ -0,0 +1,5 @@
+//
```
Comment (hkparker) [inline lib/netdb/doc.go]: Probably good to `gofmt` everything

**Comment by hkparker** [comment]
Really excited to merge this soon, interested in your thoughts on the comments.

File: lib/config/router.go (hunk 1 above)
Comment (majestrate) [inline lib/config/router.go]: I would ideally like to use java i2p's network database filesystem skiplist, i2pcpp tried sqlite and it turned out it wasn't efficient at all.

File: lib/crypto/dsa_test.go (hunk 2 above)
Comment (majestrate) [inline lib/crypto/dsa_test.go]: I have go test ./... in my pre-commit git hook, go's unit testing in the stdlib is amazing vs other language's stdlib unit testing and I see little reason to use anything else.

File: lib/crypto/ed25519.go (hunk 3 above)
Comment (majestrate) [inline lib/crypto/ed25519.go]: I would personally like to use a wrapper around libsodium to avoid any potential crypto bugs.

**Comment by majestrate** [comment]
I use [this](https://github.com/majestrate/dotfiles/blob/94b30e49e5f7f97e981de2ab85ec2ff642059674/git-hooks/go/pre-commit) in my git hooks to check the code before comitting, would suggest anyone else working on this code do the same.

**Comment by hkparker** [comment]
@majestrate I've been using [this](https://github.com/fatih/vim-go) which formats on save and then I run tests manually, should be compatible.

File: lib/crypto/dsa.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+import (
+	"crypto/dsa"
+	"crypto/rand"
+	"crypto/sha1"
+	"io"
+	"math/big"
+)
+
+var dsap = new(big.Int).SetBytes([]byte{
```
Comment (hkparker) [inline lib/crypto/dsa.go]: Where do these p,q,g values come from?

**Comment by majestrate** [comment]
http://geti2p.net/en/docs/how/cryptography#DSA

**Comment by hkparker** [comment]
Anything that would be particularly useful to work on?  Don't want to duplicate efforts.

**Comment by majestrate** [comment]
implement Lease, LeaseSet, RouterAddress, RouterInfo and Mapping from http://geti2p.net/en/docs/spec/common-structures

I'll start working on the rest of the bootstrap code during that, which means I'll be implementing RSA4096 and SU3 files.

**Comment by hkparker** [comment]
Alright, I've implemented those structures and pushed to the head of my master.  Please review the changes and let me know if there's anything I should change or any functionally to add.  If it looks good to you I'll go through and add tests, sanity checks, and comments for all the common structures.

**Comment by majestrate** [comment]
Doesn't compile, PR that fixes it inbound

**Pull request merged**